utils/message-generator/src/executor.rs

```diff
@@ -199,35 +199,21 @@ impl Executor {
                     result
                 );
 
-                // If the connection should drop at this point then let's just break the loop
-                // Can't do anything else after the connection drops.
-                if *result == ActionResult::CloseConnection {
-                    info!(
-                        "Waiting 1 sec to make sure that remote have time to close the connection"
-                    );
-                    tokio::time::sleep(std::time::Duration::from_millis(1000)).await;
-                    recv.recv()
-                        .await
-                        .expect_err("Expecting the connection to be closed: wasn't");
-                    success = true;
-                    break;
-                }
-
-                let message = match recv.recv().await {
-                    Ok(message) => message,
-                    Err(_) => {
-                        success = false;
-                        error!("Connection closed before receiving the message");
-                        break;
-                    }
-                };
-
-                let mut message: Sv2Frame<AnyMessage<'static>, _> = message.try_into().unwrap();
-                debug!("RECV {:#?}", message);
-                let header = message.get_header().unwrap();
-                let payload = message.payload();
                 match result {
                     ActionResult::MatchMessageType(message_type) => {
+                        let message = match recv.recv().await {
+                            Ok(message) => message,
+                            Err(_) => {
```

Comment:
We could use the error in the `error!` macro

Reply:
can you elaborate? I don't fully understand the suggestion

Reply:
I originally didn't understand the suggestion, but now I get where you are going, but I think you're missing some context.

If we check the [docs on `async_channel::Receiver::recv` method](https://docs.rs/async-channel/latest/async_channel/struct.Receiver.html#method.recv), we see that the only situation where this method returns an error is when the channel has been closed. And on this specific context, a closed channel can only mean one thing: the TCP connection was closed and no message was received, which is already being properly conveyed in the current error message.

If we print the actual value of `Err(e)` here we get: `receiving from an empty and closed channel`. Printing out this string into the logs wouldn't improve the UX. In fact, it would make things more confusing to the user, since there's no need for them to know that we're using `async_channel::Receiver` under the hood. With the current implementation, the user already gets all info they need to know.

I think we can leave `Err(_)` with no penalty on code quality.

Reply:
Couldnt the error explain why we are 'receiving from an empty and closed channel' ?

General q about this MG executor, is it only used in tests or  also somewhere else?

Reply:
> General q about this MG executor, is it only used in tests or also somewhere else?

The entire implementation of MG (`utils/message-generator/*`) is only used in MG tests, and no other circumstance. 

Every `*.rs` file exists with the [purpose of being called by `utils/message-generator/src/main.rs` when MG is being run](https://github.com/stratum-mining/stratum/blob/7841021e3f20280c62af119de5de58f421c6e428/utils/message-generator/src/main.rs#L419), and that includes both `utils/message-generator/src/executor.rs` + `utils/message-generator/src/executor_sv1.rs`.

There's some small nuance with regards to real tests vs mocks. Each mock means that a new MG child process (with its own executor) is created and managed by the root executor (via `cargo run`). See [this](https://github.com/stratum-mining/stratum/blob/7841021e3f20280c62af119de5de58f421c6e428/test/message-generator/test/jds-do-not-fail-on-wrong-tsdatasucc/jds-do-not-fail-on-wrong-tsdatasucc.json#L42) + [this](https://github.com/stratum-mining/stratum/blob/7841021e3f20280c62af119de5de58f421c6e428/utils/message-generator/src/executor.rs#L54) as an example.

> Couldnt the error explain why we are 'receiving from an empty and closed channel' ?

If `executor.rs` was meant to be exposed as an API for external consumption in some specific use-cases, maybe that would make sense, since we would need to bubble-up the error message to whatever code was consuming this.

That is not the case however. The only scenario where this `error!` macro provides some meaningful information is when the MG test is being executed. And the current information already provides everything the user needs to know.